### DISCUSSION TITLE
Stepper motor tuning

### DISCUSSION THREAD

**Discussion by KaruroChori** [xblax/flashforge_ad5m_klipper_mod#8] [Ideas]
At the moment most motors in the amd5 are running very hot (so much so that I am worried they are even risking to skip steps  because of that), and they are generally quite noisy, even at low speed.  
I really do not care much about getting 600mm/s with 2m/s^2, for that I have a delta printer.  

So my plan was to add a more silent profile on my adm5 :D.

The issue is that I cannot find any information about the motors used in the x-y (I have yet to check the others).
If I transcribed them correctly they should be  `bj42d22 53v02` but there is not a single datasheet for them online I can find.

The idea was to add a profile for them to https://github.com/andrewmcgr/klipper_tmc_autotune and let the extension do the heavy lifting of configuring the tmc2209.

**Comment by jlodew0** [comment]
But not any different then standard firmware I assume?
edit: I requested a datasheet from Keli motor group. (seems to be the manufacturer)
Mind you, most steppers kan take a lot of heat. More then you will bare to touch.
If I get an aswer, I will update here.
Since the values come straight out of the klipper config of FF:
[tmc2209 stepper_x]
uart_pin: PC2
interpolate:True
run_current: 0.9
#hold_current:0.3
sense_resistor:0.10
#stealthchop_threshold: 12000

[tmc2209 stepper_y]
uart_pin: PA4
interpolate:True
run_current: 0.9
#hold_current:0.3
sense_resistor:0.10
#stealthchop_threshold: 12000

[tmc2209 stepper_z]
uart_pin: PC5
interpolate:True
run_current: 0.6
hold_current:0.6
sense_resistor:0.10
#stealthchop_threshold: 999999

[tmc2209 extruder]
uart_pin: eboard:PB10
interpolate:True
run_current: 0.8
hold_current:0.8
sense_resistor:0.10
#stealthchop_threshold: 999999

**Reply by KaruroChori** [reply]
Yes yes, it is the same problem as with the original configuration.
The difference is that we have more freedom now on how to change klipper here, so I am trying to take advantage of that :D

**Comment by KaruroChori** [comment]
> edit: I requested a datasheet from Keli motor group. (seems to be the manufacturer) Mind you, most steppers kan take a lot of heat. More then you will bare to touch. If I get an aswer, I will update here. 

Thank you!

**Comment by jlodew0** [comment]
I wouldn't worry to much in the meanwhile. Most steppers are rated up to 80-100 degrees celsius.
And 50 degrees celsius already feels to hot to touch for longer dan 3 seconds. Actually 54 degrees gives you full skin thickness burn wounds after 30 seconds. (to give an idea that It may feel uncomfortable but it's ok for the stepper).
And if they eventually die after some time, they are dirt cheap these days.

**Reply by KaruroChori** [reply]
While you are right, the magnetic permeability of the core will decrease with temperature raising as well, limiting the amount of torque the motor can provide.  
And since xy motors are in the same enclosure with the rest of the electronics, I would be a bit conservative about temperatures to be honest, to avoid affecting the motherboard lifespan.

**Comment by xblax** [comment]
If we can find a tuned configuration that works reliably I'm open to changing it. Also thought about enabling stealthchop mode, but on the other hand the hotend cooling fan is so noisy it doesn't really matter, if we can't reduce the fan noise (that's a separate topic I guess).

Tuning the stepper makes sense, not only for noise, but but also for accurate extrusion.

I can check my other printer config later today, that also has Klipper with TMC2209 which I modded myself to the stock board. 

For print speeds, I think the stepper settings should work for 300mm/s printing and (maybe) 600mm/s movements. I thought about reducing the excessive acceleration a bit, but not sure what other think about it.

```
[printer]
kinematics: corexy
max_velocity: 600
max_accel: 20000
max_accel_to_decel:5000
max_z_velocity: 25
max_z_accel: 500
square_corner_velocity: 25
```

I would probably set max_accel to 10000 and max_accel_to_decel to 2500.

@KaruroChori 
You can try to manually install that Klipper extension for auto-tuning. The mod systems has no `git` at the moment also no systemd, so that auto-installer won't work.

I don't think we need to have that extension in the mod installed by default. We can just add the fine tuned stepper settings after finding them once.

**Reply by KaruroChori** [reply]
Yeah, honestly my temporary configuration is more like:
```
[printer]
kinematics: corexy
max_velocity: 150
max_accel: 5000
max_accel_to_decel:2500
max_z_velocity: 25
max_z_accel: 500
square_corner_velocity: 5
```
with stealthchop enabled, and the current on motors tuned down a bit.

So it is not possible to have a solution fitting everyone :D. But at least we could have some more reasonable default.

As for the auto tuning, I first need the datasheet for the motors. Hopefully @jlodew0 will get a reply back, otherwise I will have to directly measure those parameters myself. So for now it is paused.

**Reply by xblax** [reply]
I think you are bit to conservative :D That's close to my old "Geetech Mecreator 2" that I modded to Klipper with input shaping and TMC2209. ADM5 kinematics are much more rigid.

But I limit acceleration in slicer profiles, I have different profiles for high quality and high speed printing. In the profiles I set acceleration and speed for different features of the print. I.e. in high speed mode infill could print at 300mm/s with 10000 accel and outer perimeter with 150mm/s with 5000 accel etc. I think printer config should set reasonable upper bounds.

I believe it's a good idea to add `M204` to set lower acceleration in homing_override, i.e. to 5000 before homing. Than slicer profiles can still set higher acceleration for high speed profiles if needed.

**Reply by KaruroChori** [reply]
You probably don't have neighbors as insufferable as mine :D
It if was up to me, I would not care less about noise.

But yes, I agree. That should be tuned in the slicer, not by placing hard limits to klipper config.

**Comment by jlodew0** [comment]
Can I use stealthChop mode on an extruder with pressure advance?
(https://www.klipper3d.org/TMC_Drivers.html#can-i-use-stealthchop-mode-on-an-extruder-with-pressure-advance)
Many people successfully use "stealthChop" mode with Klipper's pressure advance. Klipper implements [smooth pressure advance](https://www.klipper3d.org/Kinematics.html#pressure-advance) which does not introduce any instantaneous velocity changes.

**However, "stealthChop" mode may produce lower motor torque and/or produce higher motor heat. It may or may not be an adequate mode for your particular printer.**

Lowering run current, decreases torque and can enventually lead to step skipping.

But in all honesty I don't think we have to worry about the temperatures. They even run on the same current in the enclosed m5 pro (so even hotter). On my ender 3 I had replaces my steppers with higher torque ones and even run them with higher current then specs stated. As for the speed, max velocity and acceleration will almost never be hit. And when slicing they get overridden most of the time by slower settings as well.

**Reply by KaruroChori** [reply]
> Can I use stealthChop mode on an extruder with pressure advance?

In my experience with other printers yes, no problem there.

**Reply by xblax** [reply]
I think on my other printer I have stealthchop disabled on the extruder stepper for that reason. Probably the micro steps should be increased and interpolate disabled.

**Reply by KaruroChori** [reply]
For reference, my other printers are not direct drive, and they all use relatively larger stepper motors, so torque is plenty. It is possible that with pancake motors having stealth-chop is a bit too close to the sun.

**Comment by jlodew0** [comment]
It seems like FF has been pretty modest with current settings:
[规格书(BJ42D22-53V02)-drawing.pdf](https://github.com/xblax/flashforge_adm5_klipper_mod/files/14402688/BJ42D22-53V02.-drawing.pdf)
p.s. I just asked and the 1.5A is RMS current.

**Reply by jlodew0** [reply]
also interesting:
TMC interpolate setting introduces small position deviation[¶](https://www.klipper3d.org/TMC_Drivers.html#tmc-interpolate-setting-introduces-small-position-deviation)
The TMC driver interpolate setting may reduce the audible noise of printer movement at the cost of introducing a small systemic positional error. This systemic positional error results from the driver's delay in executing "steps" that Klipper sends it. During constant velocity moves, this delay results in a positional error of nearly half a configured microstep (more precisely, the error is half a microstep distance minus a 512th of a full step distance). For example, on an axis with a 40mm rotation_distance, 200 steps_per_rotation, and 16 microsteps, the systemic error introduced during constant velocity moves is ~0.006mm.

For best positional accuracy consider using spreadCycle mode and disable interpolation (set interpolate: False in the TMC driver config). When configured this way, one may increase the microstep setting to reduce audible noise during stepper movement. Typically, a microstep setting of 64 or 128 will have similar audible noise as interpolation, and do so without introducing a systemic positional error.
https://www.klipper3d.org/TMC_Drivers.html

**Reply by jlodew0** [reply]
https://klipper.discourse.group/t/hold-current-and-run-current-tmc2209/1171/5
![image](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/38051653/12bb2e0a-f7d4-4996-a7e6-2cfa86b6d7fc)
https://www.analog.com/media/en/technical-documentation/data-sheets/TMC2209_datasheet_rev1.09.pdf

**Reply by xblax** [reply]
Yes the 0.9A seem to be fine. I tried lowering the current to 0.8A and enabling stealthchop mode, but the steppers did not like that all on high speed moves :D I think even on my old printer, I use 0.9A for some of the steppers.

Interpolation should be fine, because I believe that is not a cumulative error, just a tiny "lag". But we could test to increase the micro steps to 32 for a bit improved accuracy.

**Reply by KaruroChori** [reply]
> [规格书(BJ42D22-53V02)-drawing.pdf](https://github.com/xblax/flashforge_adm5_klipper_mod/files/14402688/BJ42D22-53V02.-drawing.pdf)

You got it! Sadly I will be abroad for a while, so I cannot do anything hardware-related.

**Comment by jlodew0** [comment]
> Yes the 0.9A seem to be fine. I tried lowering the current to 0.8A and enabling stealthchop mode, but the steppers did not like that all on high speed moves :D I think even on my old printer, I use 0.9A for some of the steppers.
> 
> Interpolation should be fine, because I believe that is not a cumulative error, just a tiny "lag". But we could test to increase the micro steps to 32 for a bit improved accuracy.

the fan makes so much noise that I don't see any benefit in using stealthchop.

**Reply by KaruroChori** [reply]
Which fan? If you are running the part cooling fan at 100% that clearly overshadows everything else.

I generally do not print with 100% part cooling fan. Even with PLA, it is extremely overpowered. I only raise it above 30% for bridging or on extremely  short layers.

The one cooling the hot-end is noisy but still comparable with the noise level of motors running in spread-cycles. So changing them to stealth-chop in my case has a positive impact on noise.

Also, the hot-end fan is always on while printing and at a constant level, so it is easier to "cancel it out" from the background noise. I find the same is not true for motors running in spread-cycle.

I also set the one in the electronic enclosure around 66%. It takes down a lot of noise, but the airflow seems to be still decent.

**Comment by b-fitzpatrick** [comment]
Any progress on motor tuning? They're generally OK but quite noisy with heavy VFAs at certain combinations of speed/direction.

**Reply by b-fitzpatrick** [reply]
FWIW, I've been experimenting with motor current, and they're much quieter at lower current. I'm running PETG, so not really pushing extreme speeds. Machine limits at 300 mm/s, 12000 mm/s², 6000 mm/s² A-to-D, and 5 mm/s SCV.

I kept dropping the current on the X and Y motors, and I started noticing some different sounds at 0.3 A but no layer shifts in my short test. To be safe, I bumped it back up to 0.45 A, which is half of the stock current.

I also increased microsteps from 16 to 64 and then 128, and disabled interpolate. I'll probably drop it back down to 64. Not sure how much effect this had, but Klipper docs recommend increased microsteps as alternative to interpolate.

Sound is pretty decent, now, but there's some rattle during gyroid infill. It goes away if I drop A-to-D down to 3000, but that slows it down quite a bit. I think there's too much play in the linear bearings. I have both a Pro and non-Pro. I went away from the Pro due to the motor noises seemingly being amplified by the enclosure walls, but the linear bearings seem to be tighter on it. I'll have to try the lower motor current on the Pro.

I'm new to motor tuning, so if there's anything to watch out for, please let me know.

**Reply by b-fitzpatrick** [reply]
128 microsteps are too many. Had a couple of "Timer too close" errors. Printing at 64 microsteps now.

**Reply by consp** [reply]
Interesting, might want to try that myself!

**Comment by KaruroChori** [comment]
For my future reference, to tune the registers of the tmc drivers: https://github.com/MRX8024/chopper-resonance-tuner

**Reply by KaruroChori** [reply]
The script out of the box cannot work as we are working from root and all sorts of things, but I managed to "fix" it and install the missing python dependencies (numpy which is already installed can be left to the older one).
This is a repository where I collected all changes I had to perform: https://github.com/KaruroChori/chopper-resonance-tuner-ad5m-mod

I followed the [tutorial](https://github.com/KaruroChori/chopper-resonance-tuner-ad5m-mod/blob/main/wiki/chopper_tuning_guide_english.md) and got this first graph, so everything is working as expected:
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9377545/f62eccbd-503a-4239-9585-e58f9cf14096)

Now I should run the second command to find the best parameters, but after doing it for quite a while, memory in tmp is exhausted. So more changes are needed to write those files elsewhere.

This sort of calibration can just be performed once on a single unit, and we can apply it as a default for everyone else to use, so if anyone else can run it it would be great :D. 
It is quite loud and I really don't want to push the patience of my neighbors any further for now.

**Reply by xblax** [reply]
Yes I agree on doing that calibration on a single printer, or maybe on some of ours to cross-check the results.

But what are you doing that the steppers are the loudest of you printer? I feel that the cooling still is the loudest overall, even though I have tuned down the hotend fan for my config and replaced the power supply to be mostly silent.

My neighbors luckily didn't really complain although I have run this overnight with infill at 300mm/s multiple times. I sleep in a room next to the printer (but separated by two doors) and luckily can't hear it at night.

**Reply by KaruroChori** [reply]
The second test has to be performed while matching the lowest peak in speed where resonance is highest.
So it is significantly more noisy compared to a normal printjob.

Cooling is not really that bad, and most materials aside from PLA don't want it except for overhangs. So yes, in my everyday printing, this one while in SpreadCycle is the worst offender. Even compared to other printers driven as hard in the same mode.

Aside from that, I generally use StealthChop whenever possible to further reduce noise at lower speeds.

> My neighbors luckily didn't really complain although I have run this overnight with infill at 300mm/s multiple times. I sleep in a room next to the printer (but separated by two doors) and luckily can't hear it at night.

Let's not talk about my neighbors, or I will end up getting banned from github :D.
The only printer I can keep full speed at nighttime is my modified flsun q5. That is totally silent even without any enclosure :D.
All the others are strongly tuned down if I need to continue overnight.


By the way, the graph before also means one should try to exclude 50mm/s and 100mm/s in the slicer, as those are the ones introducing more VFA.

**Reply by xblax** [reply]
> By the way, the graph before also means one should try to exclude 50mm/s and 100mm/s in the slicer, as those are the ones introducing more VFA.

That's interesting, good to know!

I tend to printer smaller parts and print speed is often limited by layer time, rather then print speed. Then I need quite a lot of cooling at least if overhangs are also included in the print.

What settings do you use for StealthChop? I enabled it once and it was louder than in SpreadCycle and had troubles with high speed (I keep it at max 300/ms). I read that voltage must be increased and that the steppers could run hotter.

**Reply by KaruroChori** [reply]
> I tend to printer smaller parts and print speed is often limited by layer time, rather then print speed. Then I need quite a lot of cooling at least if overhangs are also included in the print.

Try adding more objects on the bed :D. I usually do that to save time and keep the fan mostly off.

Out of the box it is sadly as you just said. I accept the lower speeds as a good trade off between quality, flow limitations in the material, noise and wear&tear of the machine. However it can be improved.

It is louder at high speed and the overall torque is limited, mostly because the registers are not well tuned.
Regardless, the approach suggested by Trinamic is to have a threshold to switch between the two modes. 
However, the lack of calibration of the registers is such that with the generic configuration provided in klipper might lead to some positional error. So this threshold is not usually suggested in Klipper's official guidance.

StealthChop can be hotter just because it has been poorly implemented in Klipper. In an ideal scenario it would be used alongside  CoolStep. But since its integration requires to tune the registers... both Marlin and Klipper have excluded this feature from the default configuration.

Hopefully, both tools I reported in this discussion should help mitigating these issues in slightly different ways.

**Reply by b-fitzpatrick** [reply]
Any hints on getting the pip packages installed? I was getting disk full errors, and I think I figured out that I need to remount /mnt/orig_root as writable. Now, it's just hanging and being killed (not sure why).
```
ADM5 ~ # pip install matplotlib
Collecting matplotlib
Killed
```

**Reply by KaruroChori** [reply]
Actually I have not installed it, and I was able to run it without.
I updated the readme in the modified repo to reflect this.

**Reply by xblax** [reply]
> Any hints on getting the pip packages installed? I was getting disk full errors, and I think I figured out that I need to remount /mnt/orig_root as writable. Now, it's just hanging and being killed (not sure why).

If you need to remount orig_root there's definitely something wrong. But I don't think you can install Matplotlib via pip, it needs too much resources. For know, if you want to create plots you have to do it externally.

**Reply by KaruroChori** [reply]
I checked the source, it is not needed because it uses a different library to generate the plot as html.
It is probably something which was left behind from a previous version.

**Reply by consp** [reply]
> But I don't think you can install Matplotlib via pip

Since it has libc++ dependencies (pybind11) and thus a compiler it's pretty much impossible.

**Reply by ballaswag** [reply]
A shameless plug, guppy screen has a tmc metrics panel you can enable to see your motors running current and load (sg_threshold) in a live time series. It’s something I experimented to aid in tmc tuning, but haven’t fully gotten around to doing something end to end. It looks like this https://www.reddit.com/r/klippers/s/5WpgO0lTf9

**Reply by consp** [reply]
@ballaswag  You probably already mentioned it but my memory is like a leaky bucket in reverse: what is the compile setting for enabeling the TMC statistics?

**Reply by b-fitzpatrick** [reply]
@KaruroChori I was able to run the first step and get a plot. I had to change the path to chopper_plot.py in chopper_plot.sh, as it had the "-ad5m-mod" directory. I was expecting it to move diagonally to isolate a motor, but that didn't happen. Did yours? My plot also doesn't match yours and doesn't have nice, clean peaks. Perhaps some other change I've made to printer.cfg is causing a problem.
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/c7cd9c60-84ab-4bda-9e43-1a6319f994e8)

**Reply by KaruroChori** [reply]
In a core xy the combined effect of the two motors is observed in vertical or horizontal movements.
~~Can you push those changes to the repo? Otherwise I will apply them myself.~~ changed, thanks for reporting

**Reply by KaruroChori** [reply]
Your plot is totally different form mine. Mine has a much higher magnitude on those isolate peaks.
Actually yours in a way is better. It means motors are not as bad or your config is improving something. Can you share it?

**Reply by ballaswag** [reply]
> @ballaswag  You probably already mentioned it but my memory is like a leaky bucket in reverse: what is the compile setting for enabeling the TMC statistics?

Should just be the tmc_status.py, a module that dynamically loads klippy module, and enabling them in printer.cfg.

https://github.com/ballaswag/guppyscreen/blob/8ec88ce052d209971e4aae3c0c8bf8ae0ab29213/k1/scripts/guppy_cmd.cfg#L13

Last 2 klippy modules found here:
https://github.com/ballaswag/guppyscreen/tree/main/k1/k1_mods

Some note: I only have a tmc2209, so that's all I can test, but should be work similarly for other tmc modules that has stallguard.

**Reply by b-fitzpatrick** [reply]
I understand that the combined effect of two motors would be observed for a straight X or Y movement. That's why I was expecting it to move diagonally, so it would measure a single motor. Is there a way to echo status? (i.e., I want to see what it reports for: printer.configfile.settings.printer.kinematics|string)

I'm running half of the stock motor current and higher microsteps. I've added these to printer.cfg:

```
[stepper_x]
microsteps: 64

[stepper_y]
microsteps: 64

[tmc2209 stepper_x]
run_current: 0.45
interpolate: False
stealthchop_threshold: 0

[tmc2209 stepper_y]
run_current: 0.45
interpolate: False
stealthchop_threshold: 0

[tmc2209 stepper_z]
run_current: 0.45

[tmc2209 extruder]
run_current: 0.6
```

**Reply by KaruroChori** [reply]
I think this is very much by design. If the kinematic was Cartesian, it would have used the diagonal according to the code in the script. The assumption is that the driver and motor are the same for both axis. Since those numbers are only affected by these two parameters, evaluating them combined should be fine.

Your current configuration is probably having an impact. I am not sure about the microstepping, since interpolation should work in both cases, bringing the overall result more or less to the same level. Could you perform the same test in stock conditions just to make sure there is no printer to printer variance on those frequencies?

**Reply by b-fitzpatrick** [reply]
Results are more similar in stock config (and MUCH louder).
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/49ce463a-f7d2-4485-b5f8-fb8235067d56)

**Reply by KaruroChori** [reply]
Great! So calibrate once and run everywhere should be fine.
There is no reason why the optimal values for the tmc registers should change much depending on the selected speed (within reason). Working at the resonating frequency is likely just to make their effect more visible while measuring.

But there could be dependencies with the current which are unknown, so transferring optimal registers for 0.9A to 0.45A would later require some additional validation, I think.

> and MUCH louder

I know right?! @xblax did not belive me :D.

**Reply by KaruroChori** [reply]
What we need next is to move the sampled date out of `/tmp` for collection because it cannot fit in memory, and 2h approx according to the author to fully complete the first pass of the parametric sweep.
Maybe should we save them on an usb stick, so that we can process them on the pc? I am not exactly sure that tiny cpu will cope well with such a huge dataset.

**Reply by xblax** [reply]
> I know right?! @xblax did not belive me :D.

I will just test it after you have decided on optimal values :D 

And we need a few stress test to see how the steppers behave with high speed, acceleration or bumping against the previous layer lines (no z-hop), grid infill etc.

**Reply by KaruroChori** [reply]
Yeah, that is why I was trying to outsource the 2h test needed to find them. To avoid getting outsourced from my house :D.

But I will be testing https://github.com/andrewmcgr/klipper_tmc_autotune as @jlodew0 was able to find the datasheet for the motors!

**Reply by b-fitzpatrick** [reply]
The sampled data is written to /tmp via the Klipper ACCELEROMETER_MEASURE command. There's a DATA_FOLDER constant in chopper_plot.py, but that's only for the plot operation. Right now, I'm running the test at 68 mm/s (wasn't sure if I should pick 38, 61, or 68) and just periodically moving the files to /root/tmp/. I think the plot will fail initially, but I should be able to finish moving the files over and re-run it manually. Running at 0.45A, though, if it works, I can re-run at stock currents.

**Reply by KaruroChori** [reply]
Yes, around 68mm/s was likely the best to see changes during the second phase. Good luck!

**Reply by b-fitzpatrick** [reply]
Actually, had another idea and stopped it early. I ran a test with HSTRT and HEND at the stock values of 5 and 0, respectively. So, only TBL and TOFF vary. Here's the plot with my 0.45A settings at 68 mm/s:
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/d18e133d-74e9-48a4-b932-a3376125ae03)
There's not a whole lot of variability.

Here's the same plot with stock current at 50 mm/s:
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/4c125422-0c5a-4bf9-808e-fa2e44f9e2db)

In the stock configuration, the minimum vibration is at TBL=0 and TOFF=1. I re-ran the test with those values fixed and allowed HSTRT and HEND to vary. Here' that plot:
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/ab88b98f-814b-4c4d-9686-8c483b432015)

The minimum vibration is a HSTRT=7 and HEND=7. The vibration magnitude is 3590, while it was 5955 at stock settings of 2, 3, 5, and 0 for TBL, TOFF, HSTRT, and HEND, respectively.

If someone knows how to make Klipper write the ACCELEROMETER_MEASURE data elsewhere, I can run the entire test.

**Reply by KaruroChori** [reply]
Sadly without patching it is hardcoded
https://github.com/Klipper3d/klipper/blob/434770eaf9d382cccba203a3475ace2dc0653d7b/klippy/extras/adxl345.py#L135

The good news is that it should be an easy patch for just running this test. I would prefer that to be an exposed parameters in the klipper config to be honest, but the main repo closed off all issues, so it is basically impossible to contribute.

**Reply by KaruroChori** [reply]
By the way, with such configuration your are running the motor at 100kHz which is quite above the usual cap for driving them. Ideal values should stay above the audible band and below 50kHz. Going too high can cause issues with thermal losses and radio frequency emission.
I would have selected a different configuration, maybe not as silent but more within range for a typical usage.

**Reply by b-fitzpatrick** [reply]
FWIW, here's my "improved" settings vs. stock settings in my lower-current config. It's a bit smoother with lower peaks and higher valleys.
Stock: (first plot I shared above)
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/329d2c4c-4ac0-4fa7-9743-a7fb33e41e64)

"Improved":
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/44c8ec8f-8f8d-4c25-b8da-6f7dc072b7ad)

**Reply by KaruroChori** [reply]
It would be nice to extend these measurements to speeds beyond 100mm/s to also check how these values behave during fast travel moves. But considering the baseline at 0.45A was quite low, this is still an improvement.  
You only added a bit of noise at low speeds, so I consider it a win.

**Reply by b-fitzpatrick** [reply]
>Sadly without patching it is hardcoded

So, what does "patching" entail? Temporarily replacing adx345.py with a modified copy?

**Reply by KaruroChori** [reply]
In the easiest way possible yes, and just restart klipper.
If we were to make this module more "structural" I would apply a git patch to that file during the installation of the chopper-tuner, but as a one time operation it is a degree of polish which is not worth the effort :D

Do you need assistance editing it? I am awful at python but I should be able to do that :D

**Reply by b-fitzpatrick** [reply]
No, thanks. I can handle that. Probably won't get to it today, though.

**Reply by KaruroChori** [reply]
Then I am done for today as well :D.

**Comment by b-fitzpatrick** [comment]
> By the way, with such configuration your are running the motor at 100kHz which is quite above the usual cap for driving them. Ideal values should stay above the audible band and below 50kHz. Going too high can cause issues with thermal losses and radio frequency emission. I would have selected a different configuration, maybe not as silent but more within range for a typical usage.

Thanks, I didn't know that. It looks like the settings I picked for the lower-current config are OK at 23.3 kHz.

**Reply by KaruroChori** [reply]
Yes it should be ok for most people. If you have kids particularly sensitive they might be on the edge of be annoyed by it. Some pets as well. But it is usually considered ultrasound by most accounts.

**Comment by KaruroChori** [comment]
About  https://github.com/andrewmcgr/klipper_tmc_autotune
My modified branch to work on this printer https://github.com/KaruroChori/klipper_tmc_autotune/tree/flashforge-adm5

To work, tmc.py from extras should be updated to main klipper. Ideally also the other tmc****.py should be patched with the later version.

Testing now.

**Reply by KaruroChori** [reply]
Ok I got it to work and tune the motors.
Its effectiveness should be tested, but at least it is there.
If we patch tmc.py (from the code I read it is safe to do so) we can easily distribute it alongside klipper for everyone to use in case results with it are found to be positive.

The basic configuration to append is
```
[motor_constants bjd42d22-53v02]
# Coil resistance, Ohms
resistance: 2.10
# Coil inductance, Henries
inductance: 0.0046
# Holding torque, Nm
holding_torque: 0.40
# Nominal rated current, Amps
max_current: 1.50
# Steps per revolution (1.8deg motors use 200, 0.9deg motors use 400)
steps_per_revolution: 200

[autotune_tmc stepper_x]
tuning_goal: auto
motor: bjd42d22-53v02

[autotune_tmc stepper_y]
tuning_goal: auto
motor: bjd42d22-53v02

[autotune_tmc stepper_z]
tuning_goal: auto
motor: bjd42d22-53v02
```

**Reply by KaruroChori** [reply]
I edited the inductance value on the previous post (it was mH in place of H, I cannot read). 
If anyone has copied this in their configuration file, please change it to avoid potential damage on your printer.
